File: testsuite/data/scores/governance/governance.py
from iconservice import *

TAG = 'Governance'


class SystemInterface(InterfaceScore):
    @interface
    def setRevision(self, code: int):
        pass

    @interface
    def acceptScore(self, txHash: bytes):
        pass

    @interface
    def rejectScore(self, txHash: bytes):
        pass

    @interface
    def blockScore(self, address: Address):
        pass

    @interface
    def unblockScore(self, address: Address):
        pass

    @interface
    def setStepPrice(self, price: int):
        pass

    @interface
    def setStepCost(self, type: str, cost: int):
        pass

    @interface
    def setMaxStepLimit(self, contextType: str, limit: int):
        pass

    @interface
    def addDeployer(self, address: Address):
        pass

    @interface
    def removeDeployer(self, address: Address):
        pass

    @interface
    def grantValidator(self, address: Address):
        pass

    @interface
    def revokeValidator(self, address: Address):
        pass

    @interface
    def addMember(self, address: Address):
        pass

    @interface
    def removeMember(self, address: Address):
        pass

    @interface
    def addLicense(self, contentId: str):
        pass

    @interface
    def removeLicense(self, contentId: str):
        pass

    @interface
    def setTimestampThreshold(self, threshold: int):
        pass

    @interface
    def setRoundLimitFactor(self, factor: int):
        pass

    @interface
    def setDeployerWhiteListEnabled(self, yn: bool):
        pass

    @interface
    def setMinimizeBlockGen(self, yn: bool):
        pass

    @interface
    def setUseSystemDeposit(self, address: Address, yn: bool):
        pass

    @interface
    def openBTPNetwork(self, networkTypeName: str, name: str, owner: Address):
        pass

    @interface
    def closeBTPNetwork(self, id: int):
        pass


class Governance(IconScoreBase):

    def __init__(self, db: IconScoreDatabase) -> None:
        super().__init__(db)
        self.system_score = self.create_interface_score(ZERO_SCORE_ADDRESS, SystemInterface)

    def on_install(self, name: str, value: int) -> None:
        Logger.info(f'install governance: {name}, value: {value}')
        super().on_install()

    def on_update(self) -> None:
        Logger.info('update governance')
        super().on_update()

    # NOTE: Governance should not accept incoming funds
    # @payable
    # def fallback(self):
    #     """
    #     Called when anyone sends funds to the SCORE.
    #     """
    #     Logger.debug(f'FundTransfer({self.msg.sender}, {self.msg.value})', TAG)

    @external
    def setRevision(self, code: int):
        self.system_score.setRevision(code)

    @external
    def acceptScore(self, txHash: bytes):
        self.system_score.acceptScore(txHash)

    @external
    def rejectScore(self, txHash: bytes):
        self.system_score.rejectScore(txHash)

    @external
    def blockScore(self, address: Address):
        self.system_score.blockScore(address)

    @external
    def unblockScore(self, address: Address):
        self.system_score.unblockScore(address)

    @external
    def setStepPrice(self, price: int):
        self.system_score.setStepPrice(price)

    @external
    def setStepCost(self, type: str, cost: int):
        self.system_score.setStepCost(type, cost)

    @external
    def setMaxStepLimit(self, contextType: str, limit: int):
        self.system_score.setMaxStepLimit(contextType, limit)

    @external
    def grantValidator(self, address: Address):
        self.system_score.grantValidator(address)

    @external
    def revokeValidator(self, address: Address):
        self.system_score.revokeValidator(address)

    @external
    def addMember(self, address: Address):
        self.system_score.addMember(address)

    @external
    def removeMember(self, address: Address):
        self.system_score.removeMember(address)

    @external
    def addDeployer(self, address: Address):
        self.system_score.addDeployer(address)

    @external
    def removeDeployer(self, address: Address):
        self.system_score.removeDeployer(address)

    @external
    def addLicense(self, contentId: str):
        self.system_score.addLicense(contentId)

    @external
    def removeLicense(self, contentId: str):
        self.system_score.removeLicense(contentId)

    @external
    def setTimestampThreshold(self, threshold: int):
        self.system_score.setTimestampThreshold(threshold)

    @external
    def setRoundLimitFactor(self, factor: int):
        self.system_score.setRoundLimitFactor(factor)

    @external
    def setDeployerWhiteListEnabled(self, yn: bool):
        self.system_score.setDeployerWhiteListEnabled(yn)

    @external
    def setMinimizeBlockGen(self, yn: bool):
        self.system_score.setMinimizeBlockGen(yn)

    @external
    def setUseSystemDeposit(self, address: Address, yn: bool):
        self.system_score.setUseSystemDeposit(address, yn)

    @external(readonly=True)
    def updated(self) -> bool:
        return True

    @external
    def openBTPNetwork(self, networkTypeName: str, name: str, owner: Address):
        self.system_score.openBTPNetwork(networkTypeName, name, owner)

    @external
    def closeBTPNetwork(self, id: int):
        self.system_score.closeBTPNetwork(id)

File: javaee/samples/governance/src/governance/Governance.java
/*
 * Copyright 2021 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package governance;

import score.Address;
import score.Context;
import score.annotation.External;

import java.math.BigInteger;

public class Governance {
    private final SystemInterface system;

    public Governance(String name) {
        Context.println(">>> Governance: " + name);
        system = new SystemInterface();
    }

    @External
    public void setRevision(int code) {
        system.setRevision(code);
    }

    @External
    public void setStepPrice(BigInteger price) {
        system.setStepPrice(price);
    }

    @External
    public void setStepCost(String type, BigInteger cost) {
        system.setStepCost(type, cost);
    }

    @External
    public void setMaxStepLimit(String contextType, BigInteger limit) {
        system.setMaxStepLimit(contextType, limit);
    }

    @External
    public void acceptScore(byte[] txHash) {
        system.acceptScore(txHash);
    }

    @External
    public void rejectScore(byte[] txHash) {
        system.rejectScore(txHash);
    }

    @External
    public void blockScore(Address address) {
        system.blockScore(address);
    }

    @External
    public void unblockScore(Address address) {
        system.unblockScore(address);
    }

    @External
    public void addMember(Address address) {
        system.addMember(address);
    }

    @External
    public void removeMember(Address address) {
        system.removeMember(address);
    }

    @External
    public void grantValidator(Address address) {
        system.grantValidator(address);
    }

    @External
    public void revokeValidator(Address address) {
        system.revokeValidator(address);
    }

    @External
    public void addDeployer(Address address) {
        system.addDeployer(address);
    }

    @External
    public void removeDeployer(Address address) {
        system.removeDeployer(address);
    }

    @External
    public void setDeployerWhiteListEnabled(boolean yn) {
        system.setDeployerWhiteListEnabled(yn);
    }

    @External
    public void setTimestampThreshold(int threshold) {
        system.setTimestampThreshold(threshold);
    }

    @External
    public void setRoundLimitFactor(int factor) {
        system.setRoundLimitFactor(factor);
    }

    @External
    public void setUseSystemDeposit(Address address, boolean yn) {
        system.setUseSystemDeposit(address, yn);
    }

    @External
    public BigInteger openBTPNetwork(String networkTypeName, String name, Address owner) {
        return system.openBTPMessage(networkTypeName, name, owner);
    }

    @External
    public void closeBTPNetwork(BigInteger id) {
        system.closeBTPMessage(id);
    }
}
